Throw an error for unknown args

achow101 opened the pull request:
Following #13190, gArgs is aware of all of the command line arguments. This PR has gArgs check whether the arguments provided are actually valid arguments. When an unknown argument is encountered, an error is printed to stderr and the program exist.

Since gArgs is used for everything that has command line arguments, `bitcoind`, `bitcoin-cli`, `bitcoin-qt`, `bitcoin-tx`, and `bench_bitcoin` are all effected by this change and all now have the same argument checking behavior.

Closes #1044

achow101 committed:
Give an error and exit if there are unknown parameters

If an unknown option is given via either the command line args or
the conf file, throw an error and exit

Update tests for ArgsManager knowing args

Ignore unknown options in the config file for bitcoin-cli

Fix tests and bitcoin-cli to match actual options used

achow101 committed:
Test gArgs erroring on unknown args

promag reviewed (commented):
Concept ACK.

What about known but not used arguments? For instance, `-disablewallet -changetype=bech32` should fail or nor?

MarcoFalke commented:
Ping @MeshCollider who has been working on something similar.

achow101 commented on src/util.cpp:
```diff
@@ -536,6 +565,56 @@ void ArgsManager::ForceSetArg(const std::string& strArg, const std::string& strV
 
+void ArgsManager::AddArg(const std::string& name, const std::string& help, const bool debug_only, const OptionsCategory& cat)
+{
+    m_available_args.emplace(std::pair<OptionsCategory, std::string>(cat, name), std::pair<std::string, bool>(help, debug_only));
```
Done

achow101 commented on src/util.h:
```diff
@@ -118,6 +118,31 @@ inline bool IsSwitchChar(char c)
 #endif
 }
 
+enum OptionsCategory
```
Done

achow101 commented:
> What about known but not used arguments? For instance, `-disablewallet -changetype=bech32` should fail or nor?

This will not fail since `-changetype` is a valid argument independent of `-disablewallet`. The checker currently does not consider parameter interactions. Essentially it just checks for whether an argument is part of the help text.

> Compile error due to missing override in https://github.com/achow101/bitcoin/blob/67699a9fd9a4c279c31f1d61ef44a16032739375/src/init.cpp#L79, I presume.

Fixed I think.

promag commented on src/init.cpp:
```diff
@@ -332,197 +332,183 @@ void OnRPCStopped()
+    gArgs.AddArg("-tor", "", false, OptionsCategory::HIDDEN);
+    gArgs.AddArg("-debugnet", "", false, OptionsCategory::HIDDEN);
+    gArgs.AddArg("-whitelistalwaysrelay", "", false, OptionsCategory::HIDDEN);
+    gArgs.AddArg("-prematurewitness", "", false, OptionsCategory::HIDDEN);
```
This must stay even if #13120 is merged? Ping @MarcoFalke.

MarcoFalke replied:
Nah, this should be rebased and removed after (and if) #13120 is merged

MarcoFalke replied:
I mean there is no urgent need to remove them. If they stick around for one release (marked as "deprecated, but silently ignored") that is fine.

promag reviewed (commented):
Should old arguments be in a legacy category? Otherwise launching with an outdated configuration might unnecessarily throw an error.

jnewbery commented:
Concept ACK

achow101 commented:
> Should old arguments be in a legacy category? Otherwise launching with an outdated configuration might unnecessarily throw an error.

There could be a deprecated category where a warning is given if any arguments are in that category. Then we can remove the arguments entirely later.

MeshCollider commented:
Apologies for the delay in replying, I'm very swamped with work at the moment.
It's clear that something needs to be done to fix the arguments being silently ignored, I'm fine with this approach over the one I've been working on since it is an immediate fix. If my rework would still be useful for other reasons it could still be done after this is merged anyway.
So Concept ACK from me.

MarcoFalke commented on src/util.cpp:
```diff
@@ -536,6 +565,58 @@ void ArgsManager::ForceSetArg(const std::string& strArg, const std::string& strV
+void ArgsManager::AddArg(const std::string& name, const std::string& help, const bool debug_only, const OptionsCategory& cat)
+{
+    std::pair<OptionsCategory, std::string> key(cat, name);
+    assert(m_available_args.count(key) == 0);
```
nit: I guess the key can just be the name, as two names with a different category don't make sense?

achow101 replied:
It's a hack to make std::map sort the keys by category and then alphabetically when iterating over the map so that the help text comes out correctly.

MarcoFalke replied:
I doubt it matters in practice, so I consider it a code style nit.

achow101 commented:
Rebased

MarcoFalke commented on src/chainparamsbase.h:
```diff
@@ -40,10 +40,9 @@ class CBaseChainParams
+ *Set the arguments for chainparams
  */
-void AppendParamsHelpMessages(std::string& strUsage, bool debugHelp=true);
+void SetParamsArgs();
```
For clarity I'd prefer if this was either a static class member or named appropriately, like `SetChainParamsBaseOptions`

MarcoFalke commented on src/util.cpp:
```diff
@@ -536,6 +536,56 @@ void ArgsManager::ForceSetArg(const std::string& strArg, const std::string& strV
+                usage += HelpMessageGroup(_("Register Commands:"));
+        }
+        if ((!arg.second.second || (show_debug && arg.second.second)) &&
+            (arg.first.second != "-daemon" || (arg.first.second == "-daemon" && mode == HelpMessageMode::BITCOIND))) {
```
Forcing the args manager be aware of this particular option smells wrong. Also, the boolean logic could be simplified to just `show_debug || !arg.second.second`.

For your convenience I have prepared a suitable diff that cleanly applies and simplifies the code:

<details><summary></summary>



```dif
diff --git a/src/bench/bench_bitcoin.cpp b/src/bench/bench_bitcoin.cpp
index b08d4cd253..21d23f1985 100644
--- a/src/bench/bench_bitcoin.cpp
+++ b/src/bench/bench_bitcoin.cpp
@@ -42,7 +42,7 @@ main(int argc, char** argv)
     gArgs.ParseParameters(argc, argv);
 
     if (HelpRequested(gArgs)) {
-        std::cout << gArgs.GetHelpMessage(HelpMessageMode::OTHER);
+        std::cout << gArgs.GetHelpMessage();
 
         return 0;
     }
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 8cb886cabd..3f00a95c91 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -90,7 +90,7 @@ static int AppInitRPC(int argc, char* argv[])
                   "  bitcoin-cli [options] help                " + _("List commands") + "\n" +
                   "  bitcoin-cli [options] help <command>      " + _("Get help for a command") + "\n";
 
-            strUsage += "\n" + gArgs.GetHelpMessage(HelpMessageMode::OTHER);
+            strUsage += "\n" + gArgs.GetHelpMessage();
         }
 
         fprintf(stdout, "%s", strUsage.c_str());
diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index 359bd7749a..13a4542c9c 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -95,7 +95,7 @@ static int AppInitRawTx(int argc, char* argv[])
               "  bitcoin-tx [options] <hex-tx> [commands]  " + _("Update hex-encoded bitcoin transaction") + "\n" +
               "  bitcoin-tx [options] -create [commands]   " + _("Create hex-encoded bitcoin transaction") + "\n" +
               "\n";
-        strUsage += gArgs.GetHelpMessage(HelpMessageMode::OTHER);
+        strUsage += gArgs.GetHelpMessage();
 
         fprintf(stdout, "%s", strUsage.c_str());
 
diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index 218a9e9f42..1a004ea4af 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -62,6 +62,9 @@ static bool AppInit(int argc, char* argv[])
     //
     // If Qt is used, parameters/bitcoin.conf are parsed in qt/bitcoin.cpp's main()
     SetupArgs();
+#if HAVE_DECL_DAEMON
+    gArgs.AddArg("-daemon", _("Run in the background as a daemon and accept commands"), false, OptionsCategory::OPTIONS);
+#endif
     gArgs.ParseParameters(argc, argv);
 
     // Process help and version before taking care about datadir
@@ -77,7 +80,7 @@ static bool AppInit(int argc, char* argv[])
             strUsage += "\n" + _("Usage:") + "\n" +
                   "  bitcoind [options]                     " + strprintf(_("Start %s Daemon"), _(PACKAGE_NAME)) + "\n";
 
-            strUsage += "\n" + gArgs.GetHelpMessage(HelpMessageMode::BITCOIND);
+            strUsage += "\n" + gArgs.GetHelpMessage();
         }
 
         fprintf(stdout, "%s", strUsage.c_str());
diff --git a/src/init.cpp b/src/init.cpp
index c371e13b21..cb00d58786 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -351,9 +351,6 @@ void SetupArgs()
     gArgs.AddArg("-blockreconstructionextratxn=<n>", strprintf(_("Extra transactions to keep in memory for compact block reconstructions (default: %u)"), DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN), false, OptionsCategory::OPTIONS);
     gArgs.AddArg("-blocksonly", strprintf(_("Whether to operate in a blocks only mode (default: %u)"), DEFAULT_BLOCKSONLY), true, OptionsCategory::OPTIONS);
     gArgs.AddArg("-conf=<file>", strprintf(_("Specify configuration file. Relative paths will be prefixed by datadir location. (default: %s)"), BITCOIN_CONF_FILENAME), false, OptionsCategory::OPTIONS);
-#if HAVE_DECL_DAEMON
-    gArgs.AddArg("-daemon", _("Run in the background as a daemon and accept commands"), false, OptionsCategory::OPTIONS);
-#endif
     gArgs.AddArg("-datadir=<dir>", _("Specify data directory"), false, OptionsCategory::OPTIONS);
     gArgs.AddArg("-dbbatchsize", strprintf("Maximum database write batch size in bytes (default: %u)", nDefaultDbBatchSize), true, OptionsCategory::OPTIONS);
     gArgs.AddArg("-dbcache=<n>", strprintf(_("Set database cache size in megabytes (%d to %d, default: %d)"), nMinDbCache, nMaxDbCache, nDefaultDbCache), false, OptionsCategory::OPTIONS);
diff --git a/src/interfaces/node.h b/src/interfaces/node.h
index a3f637db2d..44f53f7e78 100644
--- a/src/interfaces/node.h
+++ b/src/interfaces/node.h
@@ -7,7 +7,6 @@
 
 #include <addrdb.h>     // For banmap_t
 #include <amount.h>     // For CAmount
-#include <init.h>       // For HelpMessageMode
 #include <net.h>        // For CConnman::NumConnections
 #include <netaddress.h> // For Network
 
diff --git a/src/qt/utilitydialog.cpp b/src/qt/utilitydialog.cpp
index 79be2823a3..993d7454d6 100644
--- a/src/qt/utilitydialog.cpp
+++ b/src/qt/utilitydialog.cpp
@@ -78,7 +78,7 @@ HelpMessageDialog::HelpMessageDialog(interfaces::Node& node, QWidget *parent, bo
         cursor.insertText(header);
         cursor.insertBlock();
 
-        std::string strUsage = gArgs.GetHelpMessage(HelpMessageMode::OTHER);
+        std::string strUsage = gArgs.GetHelpMessage();
         QString coreOptions = QString::fromStdString(strUsage);
         text = version + "\n" + header + "\n" + coreOptions;
 
diff --git a/src/util.cpp b/src/util.cpp
index f59051e0e1..265f77f6ed 100644
--- a/src/util.cpp
+++ b/src/util.cpp
@@ -543,7 +543,7 @@ void ArgsManager::AddArg(const std::string& name, const std::string& help, const
     m_available_args.emplace(key, std::pair<std::string, bool>(help, debug_only));
 }
 
-std::string ArgsManager::GetHelpMessage(HelpMessageMode mode)
+std::string ArgsManager::GetHelpMessage()
 {
     const bool show_debug = gArgs.GetBoolArg("-help-debug", false);
 
@@ -578,8 +578,7 @@ std::string ArgsManager::GetHelpMessage(HelpMessageMode mode)
             else if (last_cat == OptionsCategory::REGISTER_COMMANDS)
                 usage += HelpMessageGroup(_("Register Commands:"));
         }
-        if ((!arg.second.second || (show_debug && arg.second.second)) &&
-            (arg.first.second != "-daemon" || (arg.first.second == "-daemon" && mode == HelpMessageMode::BITCOIND))) {
+        if (show_debug||!arg.second.second) {
             usage += HelpMessageOpt(arg.first.second, arg.second.first);
         }
     }
diff --git a/src/util.h b/src/util.h
index f36a9b254a..2c8fbfa312 100644
--- a/src/util.h
+++ b/src/util.h
@@ -135,12 +135,6 @@ enum class OptionsCategory
     REGISTER_COMMANDS
 };
 
-/** The help message mode determines what help message to show */
-enum class HelpMessageMode {
-    BITCOIND,
-    OTHER
-};
-
 class ArgsManager
 {
 protected:
@@ -262,7 +256,7 @@ public:
     /**
      * Get the help string
      */
-    std::string GetHelpMessage(HelpMessageMode mode);
+    std::string GetHelpMessage();
 };
 
 extern ArgsManager gArgs;
```
</details>

MarcoFalke commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -22,22 +22,27 @@ static const char* DEFAULT_PLOT_PLOTLYURL = "https://cdn.plot.ly/plotly-latest.m
 static const int64_t DEFAULT_PLOT_WIDTH = 1024;
 static const int64_t DEFAULT_PLOT_HEIGHT = 768;
 
+void SetupBenchArgs()
```
nit: Should be `static` after #13163

MarcoFalke commented on src/bitcoin-cli.cpp:
```diff
@@ -29,29 +29,26 @@ static const int DEFAULT_HTTP_CLIENT_TIMEOUT=900;
 static const int CONTINUE_EXECUTION=-1;
 
-static std::string HelpMessageCli()
+void SetupArgs()
```
Nit: More suitable name could be `static void SetupCliArgs`. Including the static after #13163

MarcoFalke commented on src/bitcoin-tx.cpp:
```diff
@@ -31,6 +31,41 @@ static bool fCreateBlank;
 static std::map<std::string,UniValue> registers;
 static const int CONTINUE_EXECUTION=-1;
 
+void SetupArgs()
```
Nit: More suitable name could be `static void SetupTxArgs`, including the `static` after #13163

MarcoFalke commented on src/init.h:
```diff
@@ -60,14 +61,11 @@ bool AppInitLockDataDirectory();
+/**
+ * Setup the arguments for gArgs
+ */
+void SetupArgs();
```
nit: A more suitable name would be `SetupServerArgs`, since this is shared between bitcoind and the gui.

MarcoFalke commented on src/qt/bitcoin.cpp:
```diff
@@ -531,6 +531,20 @@ WId BitcoinApplication::getMainWinId() const
     return window->winId();
 }
 
+void SetupUIArgs()
```
nit: Should be `static` after #13163

maflcko reviewed (commented):
utACK the first commit (eba9252867ffdb5a0e45f34eb1e281d92a0990f4).

This commit seems useful on its own and I suggest to separate it out into a new pull to aid review and hopefully decrease the need to rebase the whole pr too often.

I have left some bike shedding comments and one conceptual feedback for the first commit including a proposed diff.

*Edit: Note that this review is erroneously collapsed in GitHub's web view.*

achow101 commented:
Addressed all of @MarcoFalke's comments.

I've separated the first commit of this PR into #13190 and so this is now on top of that.

achow101 commented:
Rebased onto master following merge of #13190

MarcoFalke commented on src/util.cpp:
```diff
@@ -758,15 +789,33 @@ void ArgsManager::ReadConfigStream(std::istream& stream)
+            }
+        }
+        if (!found && !ignore_invalid_keys) {
+            throw std::runtime_error(strprintf("Invalid configuration value %s", it->string_key.c_str()));
```
Wouldn't it be more straightforward to `return false;` instead of throwing an exception here? If you need to pass a string you could pass in a non-const reference to an `std::string`.

MarcoFalke replied:
Also, why is it required in the first place to ignore invalid keys for the cli?

MarcoFalke replied:
Using an exception for control flow is just another instance of "goto", imo. Even in python (where using exceptions for control flow is considered pythonic) we had serious issues that took several years to fix up completely.
If you prefer to pass a result object with a boolean indicating success and an optional error string, I think that is fine. But using an std::runtime_error as the class for this object seems confusing and is not type safe.

MarcoFalke commented:
I haven't yet checked for myself, but just to note: `make check` should pass on each commit.

MarcoFalke commented:
Note that `git checkout b819262a5f5b9b07b0375e5573d3e66b5c555aa8  && make check` fails and breaks bisect.

achow101 commented:
Squashed 81557f4fef79bd7eb3028933dd484fa14790ed3f into b819262a5f5b9b07b0375e5573d3e66b5c555aa8 which fixes the make check failure.

MarcoFalke commented:
Note that `git checkout 0ff0ea83ef4b1f1b94988e98602adc7648d114c4 && ./test/functional/test_runner.py` fails and breaks bisect.

promag commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -33,13 +33,21 @@ static void SetupBenchArgs()
-    gArgs.ParseParameters(argc, argv);
+    std::string error;
+    if (!gArgs.ParseParameters(argc, argv, error)) {
+        fprintf(stderr,"Error parsing command line arguments: %s\n", error.c_str());
```
nit, space after `,` (8 occurrences).

MarcoFalke replied:
Note: You can install `clang-format` and run the https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy script to preempt whitespace nitpicking.

achow101 replied:
clang-formatted

promag commented on src/util.cpp:
```diff
@@ -444,6 +456,24 @@ void ArgsManager::ParseParameters(int argc, const char* const argv[])
+            if (eq_index != std::string::npos) {
+                arg_name.erase(eq_index);
+            }
+            if (arg_name == ArgsManagerHelper::GetArgWithoutNet(key) || (IsSwitchChar(key[0]) && key.size() == 1)) {
```
Maybe I'm missing something but I don't understand the second condition for 2 reasons:
 1. could be out of the loop since it doesn't depend on `arg`;
 2. `key.size() == 1` means that `key` is just `-`.

achow101 replied:
Yeah, we need to ignore a key of `-` for bitcoin-cli. I moved the check for that outside of the loop.

promag commented on src/util.cpp:
```diff
@@ -758,15 +791,35 @@ void ArgsManager::ReadConfigStream(std::istream& stream)
             m_config_args[strKey].push_back(strValue);
         }
+
+        // Check that the arg is known
```
Can you move this to a function to avoid duplicate code?

promag replied:
If you do so then I think you can inline `GetArgWithoutNet` there.

achow101 replied:
Made it into a function and inlined `GetArgWithoutNet`.

promag commented on src/util.cpp:
```diff
@@ -316,6 +316,18 @@ class ArgsManagerHelper {
+    static inline std::string GetArgWithoutNet(const std::string& key)
+    {
+        size_t option_index = key.find('.');
+        if (option_index == std::string::npos) {
```
Early return?
```cpp
if (option_index == std::string::npos) return key;
```

achow101 replied:
Done

achow101 commented on src/test/getarg_tests.cpp:
```diff
@@ -24,14 +24,24 @@ static void ResetArgs(const std::string& strArg)
     // Convert to char*:
     std::vector<const char*> vecChar;
-    for (std::string& s : vecArg)
+    for (std::string& s : vecArg) {
```
Removed unrelated change

promag commented:
Needs rebase.

achow101 commented:
Rebased

sipa commented on src/util.h:
```diff
@@ -128,6 +151,7 @@ class ArgsManager
     std::map<std::string, std::vector<std::string>> m_config_args;
     std::string m_network;
     std::set<std::string> m_network_only_args;
+    std::map<std::pair<OptionsCategory, std::string>, std::pair<std::string, bool>> m_available_args;
```
I think @promag's suggestion is better; it's more readable (struct with named arguments), and should be more efficient when looking up arguments.

Also, it can avoid duplicating the Arg objects by using `std::map<OptionsCategory, std::map<std::string, const Arg*>> m_args_by_category` instead.

achow101 replied:
@sipa bit late for that now though. The commit with that change was merged in a different PR with just that commit.

sipa replied:
@achow101 I disagree - the current data structure were sufficient for the use case in that PR (producing the help message), plus that PR did the bulk of the needs-frequent-rebase changes (the `gArgs.AddArg` calls everywhere).

I believe that a better data structure is possible for the new use case this PR introduces, and will probably simplify the PR too.

achow101 replied:
I've made some changes to the structure that are similar to what was suggested. Now I am using a single map which has a map nested inside. The key for the outer map is the category, and the value is a map for all of the args in that category. I am also using a struct to hold the help text and debug_only bool.

jonasschnelli commented on src/qt/bitcoin.cpp:
```diff
@@ -552,15 +552,9 @@ int main(int argc, char *argv[])
     // Do not refer to data directory yet, this can be overridden by Intro::pickDataDirectory
 
-    /// 2. Basic Qt initialization (not dependent on parameters or configuration)
+    /// 1. Basic Qt initialization (not dependent on parameters or configuration)
```
Are you sure this is safe?
I haven't tested, but following the code, `BitcoinApplication` constructor on L573 does use `gArgs.GetArg("-uiplatform"`.

achow101 replied:
I'm not sure about unsafe, but it certainly is not good practice to check for whether the argument is set before actually parsing the arguments. This also makes it so that the `-uiplatform` option is effectively ignored. It appears that I have been misled by the comment on this line.

How would you recommend that an error be shown to the user if they don't set the correct arguments?

sipa replied:
Checking command line arguments in a constructor generally seems to be bad practice, as it becomes pretty hard to reason about ordering. Is there a way that can be avoided (by passing in the argument value into the constructor instead?).

achow101 replied:
To get around this problem, I moved the platform style setup out of the constructor and set it up after `parseparameters`.

jonasschnelli commented:
Concept ACK

achow101 committed:
Use a struct for arguments and nested map for categories

Instead of a single map with the category and name as the key,
make m_available_args contain maps. The key will be the category and
the value is a map which actually contains the arguments for that
category. The nested map's key is the argument name, while the value
is a struct that contains the help text and whether the argument is
a debug only argument.

sipa commented on src/util.cpp:
```diff
@@ -549,48 +549,36 @@ void ArgsManager::ForceSetArg(const std::string& strArg, const std::string& strV
-            usage += HelpMessageOpt(arg.first.second, arg.second.first);
+    std::string usage = "";
+    for (const auto& arg_map : m_available_args) {
+        if (arg_map.first == OptionsCategory::OPTIONS) usage += HelpMessageGroup(_("Options:"));
```
In commit "Use a struct for arguments and nested map for categories":

This still violates the style guide; anything that has an else branch or more than a single statement in the then branch needs indentation/braces. What about using a `switch` here?

achow101 replied:
Replaced with a `switch`.

sipa commented on src/util.cpp:
```diff
@@ -456,6 +464,35 @@ void ArgsManager::ParseParameters(int argc, const char* const argv[])
+    for (const auto& arg_map : m_available_args) {
+        for (const auto& arg : arg_map.second) {
+            std::string arg_name = arg.first;
+            size_t eq_index = arg_name.find('=');
```
If you'd move this "="-splitting to `AddArg` instead, and would store the second part in `Arg` rather than in the key, this entire loop could be avoided and replaced with a map lookup.

achow101 replied:
Done.

promag commented on src/util.h:
```diff
@@ -132,22 +131,33 @@ enum class OptionsCategory
+        std::string m_help_text;
+        bool m_debug_only;
+
+        Arg(std::string help_param, std::string help_text, bool debug_only) : m_help_param(help_param), m_help_text(help_text), m_debug_only(debug_only) {};
```
2x `const std::string&`

achow101 replied:
Done

promag reviewed (commented):
LGTM, will finish review later.

promag reviewed (commented):
utACK d36728f.

achow101 commented on src/util.cpp:
```diff
@@ -549,48 +549,80 @@ void ArgsManager::ForceSetArg(const std::string& strArg, const std::string& strV
+
+    std::map<std::string, Arg>& arg_map = m_available_args[cat];
+    assert(arg_map.count(name) == 0);
+    arg_map.emplace(name.substr(0, eq_index), Arg(name.substr(eq_index, name.size() - eq_index), help, debug_only));
```
Done

sipa commented on src/util.cpp:
```diff
@@ -456,6 +464,24 @@ void ArgsManager::ParseParameters(int argc, const char* const argv[])
+
+    bool found = false;
+    for (const auto& arg_map : m_available_args) {
+        found |= arg_map.second.find(arg_no_net) != arg_map.second.end();
```
Faster: `if (arg_map.second.count(arg_no_net)) return true;`.

achow101 replied:
Done

sipa reviewed (commented):
utACK 256e2f8f332d2aa823abaec7f23f9a204a616b6d

sipa commented:
utACK 22143e4c837e89de0dc6d94943dc791124d1532b. Only difference is addressing my last comment.

laanwj commented:
This is beautiful ✨ , you almost timed this on my issue #1044's sixth birthday, why didn't I see this before.
```bash
(13112)$ src/bitcoind -esfsfjkshf
Error parsing command line arguments: Invalid parameter -esfsfjkshf
(13112)$ src/bitcoind -sdfasdf -daemon
Error parsing command line arguments: Invalid parameter -sdfasdf
(13112)$ src/bitcoind -sdfasdf=2342984 -daemon
Error parsing command line arguments: Invalid parameter -sdfasdf
(13112)$ vim ~/.bitcoin/bitcoin.conf  # introduce invalid option sfdsf
(13112)$ src/bitcoind
Error reading configuration file: Invalid configuration value sfdsf
(13112)$ DISPLAY=:10.0 src/qt/bitcoin-qt -sdfs
[gives a warning in GUI, after clicking OK it exits]
Segmentation fault (core dumped)
#0  QString::~QString ()
#1  PlatformStyle::~PlatformStyle ()
    at /home/orion/projects/bitcoin/bitcoin/src/qt/platformstyle.h:13
#2  BitcoinApplication::~BitcoinApplication (this=0x7fffffffdfb0) at /home/orion/projects/bitcoin/bitcoin/src/qt/bitcoin.cpp:357
#3  0x00005555555d9d58 in main (argc=<optimized out>, argv=<optimized out>) at /home/orion/projects/bitcoin/bitcoin/src/qt/bitcoin.cpp:760
```
So everything works as expected, except there's a segmentation fault in the GUI on early exit (not sure if this is new with this PR, but it's triggered here).

Also needs rebase after #13341.

MarcoFalke commented:
Needs rebase due to merge of #13341

laanwj commented:
tested ACK 903055730b57ae7c8d12aca2e3fd0951f12f7e9c (can confirm the gui crash is gone)

MarcoFalke commented:
utACK 903055730b57ae7c8d12aca2e3fd0951f12f7e9c

Sjors commented:
This might break one or two Docker containers that use $NETWORK=1 where that’s either “testnet” or “bitcoin”. Cc @NicolasDorier I’ve seen that in some lightning related stuff, though not sure if that was yours.

Anyway, problems like that would be immediately obvious upon the first launch after upgrading.
